problems normalizing paths causes test failures in django-oscar

Running the tests in [django-oscar](https://github.com/django-oscar/django-oscar) with v1.14.0 of appmap-python shows 4 failures:
```
FAILED tests/unit/catalogue/test_views_reviews.py::TestAddVoteView::test_voting_on_product_review_redirect_on_public_product - re.error: unbalanced parenthesis at position 71
FAILED tests/unit/catalogue/test_views_reviews.py::TestAddVoteView::test_voting_on_product_review_returns_404_on_non_public_product - re.error: unbalanced parenthesis at position 71
FAILED tests/unit/catalogue/test_views_reviews.py::TestProductReviewDetail::test_retrieving_product_review_returns_200_on_public_product - RuntimeError: No match for /catalogue/dumu03fbu03a5-title_12/reviews/70/ found with resolver <URLResolver 'tests._site.urls' (None:None) '^/'>, regex ^catalogue/\(\...
FAILED tests/unit/catalogue/test_views_reviews.py::TestProductReviewDetail::test_retrieving_product_review_returns_404_on_non_public_product - RuntimeError: No match for /catalogue/dumu03fbu03a5-title_13/reviews/71/ found with resolver <URLResolver 'tests._site.urls' (None:None) '^/'>, regex ^catalogue/\(\...
```

One of the failure stack traces:
```
ERROR    django.request:log.py:224 Internal Server Error: /catalogue/dumu03fbu03a5-title_13/reviews/71/
Traceback (most recent call last):
  File "/Users/ajp/src/django-oscar/django-oscar/.direnv/python-3.10.8/lib/python3.10/site-packages/django/core/handlers/exception.py", line 47, in inner
    response = get_response(request)
  File "/Users/ajp/src/django-oscar/django-oscar/.direnv/python-3.10.8/lib/python3.10/site-packages/appmap/django.py", line 236, in __call__
    rec, start, call_event_id = self.before_request_hook(request, request.path_info)
  File "/Users/ajp/src/django-oscar/django-oscar/.direnv/python-3.10.8/lib/python3.10/site-packages/_appmap/web_framework.py", line 170, in before_request_hook
    start, call_event_id = self.before_request_main(rec, request)
  File "/Users/ajp/src/django-oscar/django-oscar/.direnv/python-3.10.8/lib/python3.10/site-packages/appmap/django.py", line 270, in before_request_main
    normalized_path_info = normalize_path_info(req.path_info, resolved)
  File "/Users/ajp/src/django-oscar/django-oscar/.direnv/python-3.10.8/lib/python3.10/site-packages/appmap/django.py", line 195, in normalize_path_info
    regex, match = get_resolved_match(path_info, resolver)
  File "/Users/ajp/src/django-oscar/django-oscar/.direnv/python-3.10.8/lib/python3.10/site-packages/appmap/django.py", line 181, in get_resolved_match
    raise RuntimeError(
RuntimeError: No match for /catalogue/dumu03fbu03a5-title_13/reviews/71/ found with resolver <URLResolver 'tests._site.urls' (None:None) '^/'>, regex ^catalogue/\(\?P(?P<product_slug>[^/]+)\[\\w\-\]\*\)_\(\?P(?P<product_pk>[^/]+)\\d\+\)/reviews/(?P<pk>[0-9]+)/
```
`

--- a/_appmap/importer.py
+++ b/_appmap/importer.py
@@ -7,7 +7,7 @@ from collections import namedtuple
 from collections.abc import MutableSequence
 from functools import reduce
 
-import appmap.wrapt as wrapt
+from appmap import wrapt
 
 from .env import Env
 from .utils import FnType

--- a/_appmap/test/data/django/app/urls.py
+++ b/_appmap/test/data/django/app/urls.py
@@ -10,6 +10,10 @@ def user_view(_request, username):
     return django.http.HttpResponse(f"user {username}")
 
 
+def org_user_posts_view(_request, org, username):
+    return django.http.HttpResponse(f"org {org} user {username}")
+
+
 def post_view(_request, post_id):
     return django.http.HttpResponse(f"post {post_id}")
 
@@ -69,5 +73,9 @@ urlpatterns = [
     path("echo", echo_view),
     path("exception", exception_view),
     re_path(r"^post/included/", include([path("<username>", user_view)])),
+    re_path(
+        r"^(?P<org>\d+)/posts/",
+        include([path("<username>", org_user_posts_view)]),
+    ),
     re_path(r"^admincp/", include((admincp_patterns, "admin"), namespace="admin")),
 ]

--- a/_appmap/test/data/flask/app.py
+++ b/_appmap/test/data/flask/app.py
@@ -40,3 +40,8 @@ def show_post(post_id):
 def show_user_post(username, post_id):
     # Show the summary of a user's post
     return f"User {escape(username)} post {post_id}"
+
+
+@app.route("/<int:org>/posts/<username>")
+def show_org_user_posts(org, username):
+    return f"org {org} username {username}"

--- a/_appmap/test/test_command.py
+++ b/_appmap/test/test_command.py
@@ -1,8 +1,6 @@
 import json
-import os
 import re
 from distutils.dir_util import copy_tree
-from pathlib import Path
 
 import pytest
 from importlib_metadata import version

--- a/_appmap/test/test_django.py
+++ b/_appmap/test/test_django.py
@@ -12,7 +12,6 @@ import django.db
 import django.http
 import django.test
 import pytest
-from django.conf import settings
 from django.template.loader import render_to_string
 from django.test.client import MULTIPART_CONTENT
 
@@ -189,11 +188,11 @@ class TestDjangoApp:
     """
 
     @pytest.fixture(autouse=True)
-    def beforeEach(self, monkeypatch, pytester):  # pylint: disable=no-self-use
+    def beforeEach(self, monkeypatch, pytester):
         monkeypatch.setenv("PYTHONPATH", "init")
         pytester.copy_example("django")
 
-    def test_enabled(self, pytester):  # pylint: disable=no-self-use
+    def test_enabled(self, pytester):
         # To really check middleware reset, the tests must run in order,
         # so disable randomly.
         result = pytester.runpytest("-svv", "-p", "no:randomly")
@@ -206,7 +205,7 @@ class TestDjangoApp:
         events = json.loads(appmap_file.read_text())["events"]
         assert "http_server_request" in events[0]
 
-    def test_disabled(self, pytester, monkeypatch):  # pylint: disable=no-self-use
+    def test_disabled(self, pytester, monkeypatch):
         monkeypatch.setenv("APPMAP", "false")
         result = pytester.runpytest("-svv", "-p", "no:randomly", "-k", "test_request")
         result.assert_outcomes(passed=1, failed=0, errors=0)

--- a/_appmap/test/test_events.py
+++ b/_appmap/test/test_events.py
@@ -8,7 +8,6 @@ from threading import Thread
 
 import pytest
 
-import _appmap
 import appmap
 from _appmap.env import Env
 from _appmap.event import _EventIds, describe_value

--- a/_appmap/test/test_flask.py
+++ b/_appmap/test/test_flask.py
@@ -9,7 +9,7 @@ import pytest
 
 from _appmap.env import Env
 from _appmap.metadata import Metadata
-from appmap.flask import AppmapFlask, FlaskInserter
+from appmap.flask import AppmapFlask
 
 from ..test.helpers import DictIncluding
 
@@ -134,11 +134,11 @@ class TestFlaskApp:
     """
 
     @pytest.fixture(autouse=True)
-    def beforeEach(self, monkeypatch, pytester):  # pylint: disable=no-self-use
+    def beforeEach(self, monkeypatch, pytester):
         monkeypatch.setenv("PYTHONPATH", "init")
         pytester.copy_example("flask")
 
-    def test_enabled(self, pytester):  # pylint: disable=no-self-use
+    def test_enabled(self, pytester):
         result = pytester.runpytest("-svv")
 
         result.assert_outcomes(passed=1, failed=0, errors=0)
@@ -147,7 +147,7 @@ class TestFlaskApp:
         )
         assert appmap_file.exists()
 
-    def test_disabled(self, pytester, monkeypatch):  # pylint: disable=no-self-use
+    def test_disabled(self, pytester, monkeypatch):
         monkeypatch.setenv("APPMAP", "false")
 
         result = pytester.runpytest("-svv")

--- a/_appmap/test/test_metadata.py
+++ b/_appmap/test/test_metadata.py
@@ -1,8 +1,6 @@
 """Test Metadata"""
 # pylint: disable=protected-access, missing-function-docstring
 
-import pytest
-
 from _appmap.metadata import Metadata
 
 from ..test.helpers import DictIncluding

--- a/_appmap/test/test_params.py
+++ b/_appmap/test/test_params.py
@@ -7,9 +7,9 @@ import sys
 
 import pytest
 
-import vendor.wrapt.src.appmap.wrapt as wrapt
 from _appmap.event import CallEvent
 from _appmap.importer import FilterableCls, FilterableFn
+from appmap import wrapt
 
 empty_args = {"name": "args", "class": "builtins.tuple", "kind": "rest", "value": "()"}
 

--- a/_appmap/test/web_framework.py
+++ b/_appmap/test/web_framework.py
@@ -328,8 +328,8 @@ def wait_until_port_is(address, port, desired_state):
         current_state = port_state(address, port)
         if current_state == desired_state:
             break
-        else:
-            time.sleep(sleep_time)
+
+        time.sleep(sleep_time)
         count += 1
 
 

--- a/_appmap/test/web_framework.py
+++ b/_appmap/test/web_framework.py
@@ -207,6 +207,7 @@ class TestRequestCapture:
             ("/user/test_user", "/user/{username}"),
             ("/post/123", "/post/{post_id}"),
             ("/post/test_user/123/summary", "/post/{username}/{post_id}/summary"),
+            ("/123/posts/test_user", "/{org}/posts/{username}"),
         ],
     )
     def test_path_normalization(client, events, url, expected):

--- a/_appmap/utils.py
+++ b/_appmap/utils.py
@@ -32,12 +32,12 @@ class FnType(IntFlag):
     @staticmethod
     def classify(fn):
         fn_type = type(fn)
-        if fn_type == staticmethod or fn_type == types.BuiltinMethodType:
+        if fn_type in (staticmethod, types.BuiltinMethodType):
             return FnType.STATIC
-        elif fn_type == classmethod or fn_type == types.BuiltinMethodType:
+        if fn_type in (classmethod, types.BuiltinMethodType):
             return FnType.CLASS
-        else:
-            return FnType.INSTANCE
+
+        return FnType.INSTANCE
 
 
 class ThreadLocalDict(threading.local, MutableMapping):

--- a/appmap/django.py
+++ b/appmap/django.py
@@ -17,7 +17,6 @@ from django.dispatch import receiver
 from django.template import Template
 from django.urls import get_resolver, resolve
 from django.urls.exceptions import Resolver404
-from django.urls.resolvers import _route_to_regex
 
 from _appmap.env import Env
 from _appmap.event import (

--- a/appmap/django.py
+++ b/appmap/django.py
@@ -159,22 +159,15 @@ def request_params(request):
 def get_resolved_match(path_info, resolver):
     resolver_match = resolver.resolve(path_info)
 
-    # Sometimes the route returned by resolve is a regex, sometimes
-    # it's a url pattern.
-
-    # Start by assuming it's a regex.
-    regex = resolver_match.route
-
-    match = re.match(regex, path_info[1:])
-    if match:
-        return (regex, match)
-
-    # If it didn't match, it's a url pattern. Use _route_to_regex to
-    # turn it into a regex.  (url patterns sometimes start with a
-    # caret, which needs to be stripped before conversion.)
-    if regex[0] == "^":
-        regex = regex[1:]
-    regex = _route_to_regex(regex)[0]
+    # The last element of resolver_match.tried is the list of URLResolvers that matched path_info.
+    # Each URLResolver knows how to turn itself into a regex that will match the appropriate portion
+    # of the path. Iterate through them and create a complete regex for the path.
+    parts = []
+    for part in resolver_match.tried[-1]:
+        # Django inserts a leading '^'.
+        r = part.pattern.regex.pattern.lstrip("^")
+        parts.append(r)
+    regex = "".join(parts)
 
     match = re.match(regex, path_info[1:])
     if not match:

--- a/pylintrc
+++ b/pylintrc
@@ -39,7 +39,7 @@ extension-pkg-whitelist=
 fail-on=forgotten-debug-statement
 
 # Specify a score threshold under which the program will exit with error.
-fail-under=9.59
+fail-under=9.74
 
 # Interpret the stdin as a python script, whose filename needs to be passed as
 # the module_or_package argument.

--- a/pylintrc
+++ b/pylintrc
@@ -1,9 +1,36 @@
-[MASTER]
-ignore-paths= ^_appmap/test/data
+[MAIN]
+
+# Analyse import fallback blocks. This can be used to support both Python 2 and
+# 3 compatible code, which means that the block might have code that exists
+# only in one or another interpreter, leading to false positives when analysed.
+analyse-fallback-blocks=no
+
+# Clear in-memory caches upon conclusion of linting. Useful if running pylint
+# in a server-like mode.
+clear-cache-post-run=no
+
+# Load and enable all available extensions. Use --list-extensions to see a list
+# all available extensions.
+#enable-all-extensions=
+
+# In error mode, messages with a category besides ERROR or FATAL are
+# suppressed, and no reports are done by default. Error mode is compatible with
+# disabling specific errors.
+#errors-only=
+
+# Always return a 0 (non-error) status code, even if lint errors are found.
+# This is primarily useful in continuous integration scripts.
+#exit-zero=
 
 # A comma-separated list of package or module names from where C extensions may
 # be loaded. Extensions are loading into the active Python interpreter and may
 # run arbitrary code.
+extension-pkg-allow-list=
+
+# A comma-separated list of package or module names from where C extensions may
+# be loaded. Extensions are loading into the active Python interpreter and may
+# run arbitrary code. (This is an alternative name to extension-pkg-allow-list
+# for backward compatibility.)
 extension-pkg-whitelist=
 
 # Return non-zero exit code if any of these messages/categories are detected,
@@ -11,23 +38,40 @@ extension-pkg-whitelist=
 # specified are enabled, while categories only check already-enabled messages.
 fail-on=forgotten-debug-statement
 
-# Specify a score threshold to be exceeded before program exits with error.
+# Specify a score threshold under which the program will exit with error.
 fail-under=9.59
 
-# Add files or directories to the blacklist. They should be base names, not
-# paths.
+# Interpret the stdin as a python script, whose filename needs to be passed as
+# the module_or_package argument.
+#from-stdin=
+
+# Files or directories to be skipped. They should be base names, not paths.
 ignore=CVS
 
-# Add files or directories matching the regex patterns to the blacklist. The
-# regex matches against base names, not paths.
+# Add files or directories matching the regular expressions patterns to the
+# ignore-list. The regex matches against paths and can be in Posix or Windows
+# format. Because '\\' represents the directory delimiter on Windows systems,
+# it can't be used as an escape character.
+ignore-paths=^_appmap\\test\\data|^_appmap/test/data
+
+# Files or directories matching the regular expression patterns are skipped.
+# The regex matches against base names, not paths. The default value ignores
+# Emacs file locks
 ignore-patterns=
+
+# List of module names for which member attributes should not be checked
+# (useful for modules/projects where namespaces are manipulated during runtime
+# and thus existing member attributes cannot be deduced by static analysis). It
+# supports qualified module names, as well as Unix pattern matching.
+ignored-modules=
 
 # Python code to execute, usually for sys.path manipulation such as
 # pygtk.require().
 #init-hook=
 
 # Use multiple processes to speed up Pylint. Specifying 0 will auto-detect the
-# number of processors available to use.
+# number of processors available to use, and will cap the count on Windows to
+# avoid hangs.
 jobs=1
 
 # Control the amount of potential inferred values when inferring a single
@@ -42,6 +86,13 @@ load-plugins=
 # Pickle collected data for later comparisons.
 persistent=yes
 
+# Minimum Python version to use for version dependent checks. Will default to
+# the version used to run pylint.
+py-version=3.10
+
+# Discover python modules and packages in the file system subtree.
+recursive=no
+
 # When enabled, pylint would attempt to guess common misconfiguration and emit
 # user-friendly hints instead of false-positive error messages.
 suggestion-mode=yes
@@ -50,237 +101,8 @@ suggestion-mode=yes
 # active Python interpreter and may run arbitrary code.
 unsafe-load-any-extension=no
 
-
-[MESSAGES CONTROL]
-
-# Only show warnings with the listed confidence levels. Leave empty to show
-# all. Valid levels: HIGH, INFERENCE, INFERENCE_FAILURE, UNDEFINED.
-confidence=
-
-# Disable the message, report, category or checker with the given id(s). You
-# can either give multiple identifiers separated by comma (,) or put this
-# option multiple times (only on the command line, not in the configuration
-# file where it should appear only once). You can also use "--disable=all" to
-# disable everything first and then reenable specific checks. For example, if
-# you want to run only the similarities checker, you can use "--disable=all
-# --enable=similarities". If you want to run only the classes checker, but have
-# no Warning level messages displayed, use "--disable=all --enable=classes
-# --disable=W".
-disable=invalid-name, missing-function-docstring, missing-class-docstring, missing-module-docstring, consider-using-f-string
-
-# Enable the message, report, category or checker with the given id(s). You can
-# either give multiple identifier separated by comma (,) or put this option
-# multiple time (only on the command line, not in the configuration file where
-# it should appear only once). See also the "--disable" option for examples.
-enable=c-extension-no-member
-
-
-[REPORTS]
-
-# Python expression which should return a score less than or equal to 10. You
-# have access to the variables 'error', 'warning', 'refactor', and 'convention'
-# which contain the number of messages in each category, as well as 'statement'
-# which is the total number of statements analyzed. This score is used by the
-# global evaluation report (RP0004).
-evaluation=10.0 - ((float(5 * error + warning + refactor + convention) / statement) * 10)
-
-# Template used to display messages. This is a python new-style format string
-# used to format the message information. See doc for all details.
-#msg-template=
-
-# Set the output format. Available formats are text, parseable, colorized, json
-# and msvs (visual studio). You can also give a reporter class, e.g.
-# mypackage.mymodule.MyReporterClass.
-output-format=text
-
-# Tells whether to display a full report or only the messages.
-reports=no
-
-# Activate the evaluation score.
-score=yes
-
-
-[REFACTORING]
-
-# Maximum number of nested blocks for function / method body
-max-nested-blocks=5
-
-# Complete name of functions that never returns. When checking for
-# inconsistent-return-statements if a never returning function is called then
-# it will be considered as an explicit return statement and no message will be
-# printed.
-never-returning-functions=sys.exit
-
-
-[LOGGING]
-
-# The type of string formatting that logging methods do. `old` means using %
-# formatting, `new` is for `{}` formatting.
-logging-format-style=old
-
-# Logging modules to check that the string format arguments are in logging
-# function parameter format.
-logging-modules=logging
-
-
-[SPELLING]
-
-# Limits count of emitted suggestions for spelling mistakes.
-max-spelling-suggestions=4
-
-# Spelling dictionary name. Available dictionaries: none. To make it work,
-# install the python-enchant package.
-spelling-dict=
-
-# List of comma separated words that should not be checked.
-spelling-ignore-words=
-
-# A path to a file that contains the private dictionary; one word per line.
-spelling-private-dict-file=
-
-# Tells whether to store unknown words to the private dictionary (see the
-# --spelling-private-dict-file option) instead of raising a message.
-spelling-store-unknown-words=no
-
-
-[MISCELLANEOUS]
-
-# List of note tags to take in consideration, separated by a comma.
-notes=FIXME,
-      TODO
-
-# Regular expression of note tags to take in consideration.
-#notes-rgx=
-
-
-[TYPECHECK]
-
-# List of decorators that produce context managers, such as
-# contextlib.contextmanager. Add to this list to register other decorators that
-# produce valid context managers.
-contextmanager-decorators=contextlib.contextmanager
-
-# List of members which are set dynamically and missed by pylint inference
-# system, and so shouldn't trigger E1101 when accessed. Python regular
-# expressions are accepted.
-generated-members=
-
-# Tells whether missing members accessed in mixin class should be ignored. A
-# mixin class is detected if its name ends with "mixin" (case insensitive).
-ignore-mixin-members=yes
-
-# Tells whether to warn about missing members when the owner of the attribute
-# is inferred to be None.
-ignore-none=yes
-
-# This flag controls whether pylint should warn about no-member and similar
-# checks whenever an opaque object is returned when inferring. The inference
-# can return multiple potential results while evaluating a Python object, but
-# some branches might not be evaluated, which results in partial inference. In
-# that case, it might be useful to still emit no-member and other checks for
-# the rest of the inferred objects.
-ignore-on-opaque-inference=yes
-
-# List of class names for which member attributes should not be checked (useful
-# for classes with dynamically set attributes). This supports the use of
-# qualified names.
-ignored-classes=optparse.Values,thread._local,_thread._local
-
-# List of module names for which member attributes should not be checked
-# (useful for modules/projects where namespaces are manipulated during runtime
-# and thus existing member attributes cannot be deduced by static analysis). It
-# supports qualified module names, as well as Unix pattern matching.
-ignored-modules=
-
-# Show a hint with possible names when a member name was not found. The aspect
-# of finding the hint is based on edit distance.
-missing-member-hint=yes
-
-# The minimum edit distance a name should have in order to be considered a
-# similar match for a missing member name.
-missing-member-hint-distance=1
-
-# The total number of similar names that should be taken in consideration when
-# showing a hint for a missing member.
-missing-member-max-choices=1
-
-# List of decorators that change the signature of a decorated function.
-signature-mutators=
-
-
-[VARIABLES]
-
-# List of additional names supposed to be defined in builtins. Remember that
-# you should avoid defining new builtins when possible.
-additional-builtins=
-
-# Tells whether unused global variables should be treated as a violation.
-allow-global-unused-variables=yes
-
-# List of strings which can identify a callback function by name. A callback
-# name must start or end with one of those strings.
-callbacks=cb_,
-          _cb
-
-# A regular expression matching the name of dummy variables (i.e. expected to
-# not be used).
-dummy-variables-rgx=_+$|(_[a-zA-Z0-9_]*[a-zA-Z0-9]+?$)|dummy|^ignored_|^unused_
-
-# Argument names that match this expression will be ignored. Default to name
-# with leading underscore.
-ignored-argument-names=_.*|^ignored_|^unused_
-
-# Tells whether we should check for unused import in __init__ files.
-init-import=no
-
-# List of qualified module names which can have objects that can redefine
-# builtins.
-redefining-builtins-modules=six.moves,past.builtins,future.builtins,builtins,io
-
-
-[FORMAT]
-
-# Expected format of line ending, e.g. empty (any line ending), LF or CRLF.
-expected-line-ending-format=
-
-# Regexp for a line that is allowed to be longer than the limit.
-ignore-long-lines=^\s*(# )?<?https?://\S+>?$
-
-# Number of spaces of indent required inside a hanging or continued line.
-indent-after-paren=4
-
-# String used as indentation unit. This is usually "    " (4 spaces) or "\t" (1
-# tab).
-indent-string='    '
-
-# Maximum number of characters on a single line.
-max-line-length=100
-
-# Maximum number of lines in a module.
-max-module-lines=1000
-
-# Allow the body of a class to be on the same line as the declaration if body
-# contains single statement.
-single-line-class-stmt=no
-
-# Allow the body of an if to be on the same line as the test if there is no
-# else.
-single-line-if-stmt=no
-
-
-[SIMILARITIES]
-
-# Ignore comments when computing similarities.
-ignore-comments=yes
-
-# Ignore docstrings when computing similarities.
-ignore-docstrings=yes
-
-# Ignore imports when computing similarities.
-ignore-imports=no
-
-# Minimum lines number of a similarity.
-min-similarity-lines=4
+# In verbose mode, extra non-checker-related info will be displayed.
+#verbose=
 
 
 [BASIC]
@@ -289,13 +111,15 @@ min-similarity-lines=4
 argument-naming-style=snake_case
 
 # Regular expression matching correct argument names. Overrides argument-
-# naming-style.
+# naming-style. If left empty, argument names will be checked with the set
+# naming style.
 #argument-rgx=
 
 # Naming style matching correct attribute names.
 attr-naming-style=snake_case
 
 # Regular expression matching correct attribute names. Overrides attr-naming-
+# style. If left empty, attribute names will be checked with the set naming
 # style.
 #attr-rgx=
 
@@ -315,20 +139,30 @@ bad-names-rgxs=
 class-attribute-naming-style=any
 
 # Regular expression matching correct class attribute names. Overrides class-
-# attribute-naming-style.
+# attribute-naming-style. If left empty, class attribute names will be checked
+# with the set naming style.
 #class-attribute-rgx=
+
+# Naming style matching correct class constant names.
+class-const-naming-style=UPPER_CASE
+
+# Regular expression matching correct class constant names. Overrides class-
+# const-naming-style. If left empty, class constant names will be checked with
+# the set naming style.
+#class-const-rgx=
 
 # Naming style matching correct class names.
 class-naming-style=PascalCase
 
 # Regular expression matching correct class names. Overrides class-naming-
-# style.
+# style. If left empty, class names will be checked with the set naming style.
 #class-rgx=
 
 # Naming style matching correct constant names.
 const-naming-style=UPPER_CASE
 
 # Regular expression matching correct constant names. Overrides const-naming-
+# style. If left empty, constant names will be checked with the set naming
 # style.
 #const-rgx=
 
@@ -340,7 +174,8 @@ docstring-min-length=-1
 function-naming-style=snake_case
 
 # Regular expression matching correct function names. Overrides function-
-# naming-style.
+# naming-style. If left empty, function names will be checked with the set
+# naming style.
 #function-rgx=
 
 # Good variable names which should always be accepted, separated by a comma.
@@ -363,21 +198,22 @@ include-naming-hint=no
 inlinevar-naming-style=any
 
 # Regular expression matching correct inline iteration names. Overrides
-# inlinevar-naming-style.
+# inlinevar-naming-style. If left empty, inline iteration names will be checked
+# with the set naming style.
 #inlinevar-rgx=
 
 # Naming style matching correct method names.
 method-naming-style=snake_case
 
 # Regular expression matching correct method names. Overrides method-naming-
-# style.
+# style. If left empty, method names will be checked with the set naming style.
 #method-rgx=
 
 # Naming style matching correct module names.
 module-naming-style=snake_case
 
 # Regular expression matching correct module names. Overrides module-naming-
-# style.
+# style. If left empty, module names will be checked with the set naming style.
 #module-rgx=
 
 # Colon-delimited sets of names that determine each other's naming style when
@@ -393,66 +229,23 @@ no-docstring-rgx=^_
 # These decorators are taken in consideration only for invalid-name.
 property-classes=abc.abstractproperty
 
+# Regular expression matching correct type variable names. If left empty, type
+# variable names will be checked with the set naming style.
+#typevar-rgx=
+
 # Naming style matching correct variable names.
 variable-naming-style=snake_case
 
 # Regular expression matching correct variable names. Overrides variable-
-# naming-style.
+# naming-style. If left empty, variable names will be checked with the set
+# naming style.
 #variable-rgx=
 
 
-[STRING]
-
-# This flag controls whether inconsistent-quotes generates a warning when the
-# character used as a quote delimiter is used inconsistently within a module.
-check-quote-consistency=no
-
-# This flag controls whether the implicit-str-concat should generate a warning
-# on implicit string concatenation in sequences defined over several lines.
-check-str-concat-over-line-jumps=no
-
-
-[IMPORTS]
-
-# List of modules that can be imported at any level, not just the top level
-# one.
-allow-any-import-level=
-
-# Allow wildcard imports from modules that define __all__.
-allow-wildcard-with-all=no
-
-# Analyse import fallback blocks. This can be used to support both Python 2 and
-# 3 compatible code, which means that the block might have code that exists
-# only in one or another interpreter, leading to false positives when analysed.
-analyse-fallback-blocks=no
-
-# Deprecated modules which should not be used, separated by a comma.
-deprecated-modules=optparse,tkinter.tix
-
-# Create a graph of external dependencies in the given file (report RP0402 must
-# not be disabled).
-ext-import-graph=
-
-# Create a graph of every (i.e. internal and external) dependencies in the
-# given file (report RP0402 must not be disabled).
-import-graph=
-
-# Create a graph of internal dependencies in the given file (report RP0402 must
-# not be disabled).
-int-import-graph=
-
-# Force import order to recognize a module as part of the standard
-# compatibility libraries.
-known-standard-library=
-
-# Force import order to recognize a module as part of a third party library.
-known-third-party=enchant
-
-# Couples of modules and preferred modules, separated by a comma.
-preferred-modules=
-
-
 [CLASSES]
+
+# Warn about protected attribute access inside special methods
+check-protected-access-in-special-methods=no
 
 # List of method names used to declare (i.e. assign) instance attributes.
 defining-attr-methods=__init__,
@@ -476,6 +269,14 @@ valid-metaclass-classmethod-first-arg=cls
 
 
 [DESIGN]
+
+# List of regular expressions of class ancestor names to ignore when counting
+# public methods (see R0903)
+exclude-too-few-public-methods=
+
+# List of qualified class names to ignore when counting class parents (see
+# R0901)
+ignored-parents=
 
 # Maximum number of arguments for function / method.
 max-args=5
@@ -510,7 +311,320 @@ min-public-methods=2
 
 [EXCEPTIONS]
 
-# Exceptions that will emit a warning when being caught. Defaults to
-# "BaseException, Exception".
-overgeneral-exceptions=BaseException,
-                       Exception
+# Exceptions that will emit a warning when caught.
+overgeneral-exceptions=builtins.BaseException,
+                       builtins.Exception
+
+
+[FORMAT]
+
+# Expected format of line ending, e.g. empty (any line ending), LF or CRLF.
+expected-line-ending-format=
+
+# Regexp for a line that is allowed to be longer than the limit.
+ignore-long-lines=^\s*(# )?<?https?://\S+>?$
+
+# Number of spaces of indent required inside a hanging or continued line.
+indent-after-paren=4
+
+# String used as indentation unit. This is usually "    " (4 spaces) or "\t" (1
+# tab).
+indent-string='    '
+
+# Maximum number of characters on a single line.
+max-line-length=100
+
+# Maximum number of lines in a module.
+max-module-lines=1000
+
+# Allow the body of a class to be on the same line as the declaration if body
+# contains single statement.
+single-line-class-stmt=no
+
+# Allow the body of an if to be on the same line as the test if there is no
+# else.
+single-line-if-stmt=no
+
+
+[IMPORTS]
+
+# List of modules that can be imported at any level, not just the top level
+# one.
+allow-any-import-level=
+
+# Allow explicit reexports by alias from a package __init__.
+allow-reexport-from-package=no
+
+# Allow wildcard imports from modules that define __all__.
+allow-wildcard-with-all=no
+
+# Deprecated modules which should not be used, separated by a comma.
+deprecated-modules=optparse,tkinter.tix
+
+# Output a graph (.gv or any supported image format) of external dependencies
+# to the given file (report RP0402 must not be disabled).
+ext-import-graph=
+
+# Output a graph (.gv or any supported image format) of all (i.e. internal and
+# external) dependencies to the given file (report RP0402 must not be
+# disabled).
+import-graph=
+
+# Output a graph (.gv or any supported image format) of internal dependencies
+# to the given file (report RP0402 must not be disabled).
+int-import-graph=
+
+# Force import order to recognize a module as part of the standard
+# compatibility libraries.
+known-standard-library=
+
+# Force import order to recognize a module as part of a third party library.
+known-third-party=enchant
+
+# Couples of modules and preferred modules, separated by a comma.
+preferred-modules=
+
+
+[LOGGING]
+
+# The type of string formatting that logging methods do. `old` means using %
+# formatting, `new` is for `{}` formatting.
+logging-format-style=old
+
+# Logging modules to check that the string format arguments are in logging
+# function parameter format.
+logging-modules=logging
+
+
+[MESSAGES CONTROL]
+
+# Only show warnings with the listed confidence levels. Leave empty to show
+# all. Valid levels: HIGH, CONTROL_FLOW, INFERENCE, INFERENCE_FAILURE,
+# UNDEFINED.
+confidence=HIGH,
+           CONTROL_FLOW,
+           INFERENCE,
+           INFERENCE_FAILURE,
+           UNDEFINED
+
+# Disable the message, report, category or checker with the given id(s). You
+# can either give multiple identifiers separated by comma (,) or put this
+# option multiple times (only on the command line, not in the configuration
+# file where it should appear only once). You can also use "--disable=all" to
+# disable everything first and then re-enable specific checks. For example, if
+# you want to run only the similarities checker, you can use "--disable=all
+# --enable=similarities". If you want to run only the classes checker, but have
+# no Warning level messages displayed, use "--disable=all --enable=classes
+# --disable=W".
+disable=raw-checker-failed,
+        bad-inline-option,
+        locally-disabled,
+        file-ignored,
+        suppressed-message,
+        useless-suppression,
+        deprecated-pragma,
+        use-symbolic-message-instead,
+        invalid-name,
+        missing-function-docstring,
+        missing-class-docstring,
+        missing-module-docstring,
+        consider-using-f-string
+
+# Enable the message, report, category or checker with the given id(s). You can
+# either give multiple identifier separated by comma (,) or put this option
+# multiple time (only on the command line, not in the configuration file where
+# it should appear only once). See also the "--disable" option for examples.
+enable=c-extension-no-member
+
+
+[METHOD_ARGS]
+
+# List of qualified names (i.e., library.method) which require a timeout
+# parameter e.g. 'requests.api.get,requests.api.post'
+timeout-methods=requests.api.delete,requests.api.get,requests.api.head,requests.api.options,requests.api.patch,requests.api.post,requests.api.put,requests.api.request
+
+
+[MISCELLANEOUS]
+
+# List of note tags to take in consideration, separated by a comma.
+notes=FIXME,
+      TODO
+
+# Regular expression of note tags to take in consideration.
+notes-rgx=
+
+
+[REFACTORING]
+
+# Maximum number of nested blocks for function / method body
+max-nested-blocks=5
+
+# Complete name of functions that never returns. When checking for
+# inconsistent-return-statements if a never returning function is called then
+# it will be considered as an explicit return statement and no message will be
+# printed.
+never-returning-functions=sys.exit
+
+
+[REPORTS]
+
+# Python expression which should return a score less than or equal to 10. You
+# have access to the variables 'fatal', 'error', 'warning', 'refactor',
+# 'convention', and 'info' which contain the number of messages in each
+# category, as well as 'statement' which is the total number of statements
+# analyzed. This score is used by the global evaluation report (RP0004).
+evaluation=10.0 - ((float(5 * error + warning + refactor + convention) / statement) * 10)
+
+# Template used to display messages. This is a python new-style format string
+# used to format the message information. See doc for all details.
+msg-template=
+
+# Set the output format. Available formats are text, parseable, colorized, json
+# and msvs (visual studio). You can also give a reporter class, e.g.
+# mypackage.mymodule.MyReporterClass.
+#output-format=
+
+# Tells whether to display a full report or only the messages.
+reports=no
+
+# Activate the evaluation score.
+score=yes
+
+
+[SIMILARITIES]
+
+# Comments are removed from the similarity computation
+ignore-comments=yes
+
+# Docstrings are removed from the similarity computation
+ignore-docstrings=yes
+
+# Imports are removed from the similarity computation
+ignore-imports=no
+
+# Signatures are removed from the similarity computation
+ignore-signatures=yes
+
+# Minimum lines number of a similarity.
+min-similarity-lines=4
+
+
+[SPELLING]
+
+# Limits count of emitted suggestions for spelling mistakes.
+max-spelling-suggestions=4
+
+# Spelling dictionary name. Available dictionaries: none. To make it work,
+# install the 'python-enchant' package.
+spelling-dict=
+
+# List of comma separated words that should be considered directives if they
+# appear at the beginning of a comment and should not be checked.
+spelling-ignore-comment-directives=fmt: on,fmt: off,noqa:,noqa,nosec,isort:skip,mypy:
+
+# List of comma separated words that should not be checked.
+spelling-ignore-words=
+
+# A path to a file that contains the private dictionary; one word per line.
+spelling-private-dict-file=
+
+# Tells whether to store unknown words to the private dictionary (see the
+# --spelling-private-dict-file option) instead of raising a message.
+spelling-store-unknown-words=no
+
+
+[STRING]
+
+# This flag controls whether inconsistent-quotes generates a warning when the
+# character used as a quote delimiter is used inconsistently within a module.
+check-quote-consistency=no
+
+# This flag controls whether the implicit-str-concat should generate a warning
+# on implicit string concatenation in sequences defined over several lines.
+check-str-concat-over-line-jumps=no
+
+
+[TYPECHECK]
+
+# List of decorators that produce context managers, such as
+# contextlib.contextmanager. Add to this list to register other decorators that
+# produce valid context managers.
+contextmanager-decorators=contextlib.contextmanager
+
+# List of members which are set dynamically and missed by pylint inference
+# system, and so shouldn't trigger E1101 when accessed. Python regular
+# expressions are accepted.
+generated-members=
+
+# Tells whether to warn about missing members when the owner of the attribute
+# is inferred to be None.
+ignore-none=yes
+
+# This flag controls whether pylint should warn about no-member and similar
+# checks whenever an opaque object is returned when inferring. The inference
+# can return multiple potential results while evaluating a Python object, but
+# some branches might not be evaluated, which results in partial inference. In
+# that case, it might be useful to still emit no-member and other checks for
+# the rest of the inferred objects.
+ignore-on-opaque-inference=yes
+
+# List of symbolic message names to ignore for Mixin members.
+ignored-checks-for-mixins=no-member,
+                          not-async-context-manager,
+                          not-context-manager,
+                          attribute-defined-outside-init
+
+# List of class names for which member attributes should not be checked (useful
+# for classes with dynamically set attributes). This supports the use of
+# qualified names.
+ignored-classes=optparse.Values,thread._local,_thread._local
+
+# Show a hint with possible names when a member name was not found. The aspect
+# of finding the hint is based on edit distance.
+missing-member-hint=yes
+
+# The minimum edit distance a name should have in order to be considered a
+# similar match for a missing member name.
+missing-member-hint-distance=1
+
+# The total number of similar names that should be taken in consideration when
+# showing a hint for a missing member.
+missing-member-max-choices=1
+
+# Regex pattern to define which classes are considered mixins.
+mixin-class-rgx=.*[Mm]ixin
+
+# List of decorators that change the signature of a decorated function.
+signature-mutators=
+
+
+[VARIABLES]
+
+# List of additional names supposed to be defined in builtins. Remember that
+# you should avoid defining new builtins when possible.
+additional-builtins=
+
+# Tells whether unused global variables should be treated as a violation.
+allow-global-unused-variables=yes
+
+# List of names allowed to shadow builtins
+allowed-redefined-builtins=
+
+# List of strings which can identify a callback function by name. A callback
+# name must start or end with one of those strings.
+callbacks=cb_,
+          _cb
+
+# A regular expression matching the name of dummy variables (i.e. expected to
+# not be used).
+dummy-variables-rgx=_+$|(_[a-zA-Z0-9_]*[a-zA-Z0-9]+?$)|dummy|^ignored_|^unused_
+
+# Argument names that match this expression will be ignored.
+ignored-argument-names=_.*|^ignored_|^unused_
+
+# Tells whether we should check for unused import in __init__ files.
+init-import=no
+
+# List of qualified module names which can have objects that can redefine
+# builtins.
+redefining-builtins-modules=six.moves,past.builtins,future.builtins,builtins,io

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -13,8 +13,7 @@ license = "MIT"
 classifiers = [
         'Development Status :: 4 - Beta',
         'Framework :: Django',
-        'Framework :: Django :: 2.2',
-        'Framework :: Django :: 3.1',
+        'Framework :: Django :: 3.2',
         'Framework :: Flask',
         'Framework :: Pytest',
         'Intended Audience :: Developers',
@@ -32,21 +31,18 @@ packages = [
 [tool.poetry.dependencies]
 # Please update the documentation if changing the supported python version
 # https://github.com/applandinc/applandinc.github.io/blob/master/_docs/reference/appmap-python.md#supported-versions
-python = "^3.7"
+python = "^3.7.2"
 PyYAML = ">=5.3.0"
 inflection = ">=0.3.0"
 importlib-metadata = ">=0.8"
 importlib-resources = "^5.4.0"
 packaging = "^21.3"
-# If you include "Django" as an optional dependency here, you'll be able to use
-# poetry to install it in your dev environment. However, doing so causes poetry
-# v1.2.0 to remove it from the virtualenv *created and managed by tox*, i.e. not
-# your dev environment. 
+# If you include "Django" as an optional dependency here, you'll be able to use poetry to install it
+# in your dev environment. However, doing so causes poetry v1.2.0 to remove it from the virtualenv
+# *created and managed by tox*, i.e. not your dev environment. 
 # 
-# So, if you'd like to run the tests outside of tox,
-# you'll need to `pip install` the appropriate version of Django, e.g.
-#   pip install django'>=3.2,<4.0'
-#   pip install pytest-django
+# So, if you'd like to run the tests outside of tox, run `pip install -r requirements-dev.txt` to
+# install it and the rest of the dev dependencies.
 
 [tool.poetry.dev-dependencies]
 httpretty = "^1.0.5"
@@ -67,6 +63,7 @@ python-decouple = "^3.5"
 [tool.poetry.group.dev.dependencies]
 black = "^22.8.0"
 isort = "^5.10.1"
+pylint = "^2.16.2"
 
 [build-system]
 requires = ["poetry-core>=1.1.0"]

--- a/tox.ini
+++ b/tox.ini
@@ -2,7 +2,7 @@
 isolated_build = true
 # The *-web environments test the latest versions of Django and Flask with the full test suite. For
 # older version of the web frameworks, just run the tests that are specific to them.
-envlist = py3{8,9,10}-web, py3{7,8,9,10}-flask1-django{2,3}
+envlist = py3{8,9,10}-web, py3{7,8,9,10}-flask1-django3
 
 [testenv]
 
@@ -12,7 +12,6 @@ deps=
     web: Django >=4.0, <5.0
     web: Flask >= 2
     flask1: -rrequirements-flask1.txt
-    django2: Django >=2.2, <3.0
     django3: Django >=3.2, <4.0
 
     
@@ -24,5 +23,4 @@ commands =
     py310-web: poetry run pylint -j 0 appmap _appmap
     web: poetry run {posargs:pytest -vv}
     flask1: poetry run pytest _appmap/test/test_flask.py
-    django2: poetry run pytest _appmap/test/test_django.py
     django3: poetry run pytest _appmap/test/test_django.py

